src/pages/api/webhooks/klarna/authorization.ts

```diff
@@ -0,0 +1,166 @@
+    getConfigurationForChannel(appConfig, channelId),
+  );
+
+  const klarnaClient = getKlarnaApiClient({
+    klarnaApiUrl: klarnaConfig.apiUrl,
+    username: klarnaConfig.username,
+    password: klarnaConfig.password,
+  });
+
+  const createKlarnaOrder = klarnaClient
+    .path("/payments/v1/authorizations/{authorizationToken}/order")
+    .method("post")
+    .create();
+
+  const sourceObject =
+    transaction.data?.transaction?.checkout ?? transaction.data?.transaction?.order;
+
+  if (!sourceObject) {
+    logger.debug(transaction.data);
+    logger.error({ transactionId, channelId, saleorApiUrl }, "Not found");
+    return res.status(404).json("[NOT FOUND]");
+  }
+
+  const country = sourceObject.billingAddress?.country.code;
+  invariant(country, "Missing country code");
+
+  const email = sourceObject.userEmail;
+  invariant(email, "Missing email");
+
+  const metadata: KlarnaMetadata = {
+    transactionId,
+    channelId,
+    ...(sourceObject.__typename === "Checkout" && { checkoutId: sourceObject.id }),
+    ...(sourceObject.__typename === "Order" && { orderId: sourceObject.id }),
+  };
+
+  const createKlarnaOrderPayload: components["schemas"]["create_order_request"] = {
+    purchase_country: country,
+    purchase_currency: sourceObject.total.gross.currency,
+    billing_address: prepareRequestAddress(sourceObject.billingAddress, email),
+    shipping_address: prepareRequestAddress(sourceObject.shippingAddress, email),
+    order_amount: getKlarnaIntegerAmountFromSaleor(
+      sourceObject.total.gross.amount,
+      sourceObject.total.gross.currency,
+    ),
+    order_tax_amount: getKlarnaIntegerAmountFromSaleor(
+      sourceObject.total.tax.amount,
+      sourceObject.total.tax.currency,
+    ),
+    order_lines: getLineItems(sourceObject),
+    auto_capture: true,
+    merchant_reference1: sourceObject.id,
+    merchant_reference2: transactionId,
+    merchant_data: JSON.stringify(metadata),
+  };
+  logger.debug({ ...obfuscateConfig(createKlarnaOrderPayload) }, "createKlarnaOrder payload");
+
+  const klarnaOrder = await createKlarnaOrder({
+    ...createKlarnaOrderPayload,
+    authorizationToken,
+  });
+  logger.debug({ ...obfuscateConfig(klarnaOrder) }, "createKlarnaOrder result");
+
+  if (!klarnaOrder.ok) {
+    throw new KlarnaHttpClientError(klarnaOrder.statusText, { errors: [klarnaOrder.data] });
+  }
+
+  await client
+    .mutation(TransactionEventReportDocument, {
+      transactionId,
+      amount: sourceObject.total.gross.amount,
+
+      availableActions:
+        klarnaOrder.data.fraud_status === "ACCEPTED"
+          ? [TransactionActionEnum.Refund]
+          : [TransactionActionEnum.Cancel],
+      externalUrl: "",
+      time: new Date().toISOString(),
+      type:
+        klarnaOrder.data.fraud_status === "ACCEPTED"
+          ? TransactionEventTypeEnum.ChargeSuccess
+          : TransactionEventTypeEnum.ChargeRequest,
+      pspReference: klarnaOrder.data.order_id,
```

Comment:
Is there a way to also include klarna_reference? https://docs.klarna.com/api/ordermanagement/#operation/getOrder!c=200&path=klarna_reference&t=response

It seems to be used with customer communication so staff users will try to look for this

Reply:
We could add that if that's necessary. However, it'll require one more roundabout to the Klarna server to fetch because the order creation endpoint doesn't return it: https://docs.klarna.com/api/payments/#operation/createOrder

Reply:
Let's skip it for now 👍 . Added task for investigating it's purpose and consulting with the client